### PR TITLE
✨ Aliquot resource

### PR DESCRIPTION
**Pull request by znatty22**
Add aliquot resource with crud methods and unit tests

* Note that nested de/serialization does not work right now. For example, cannot create/update/read a sample with embedded aliquots. Also cannot create/update/read aliquots with embedded sequencing experiments. This functionality will implemented in near future.

### CODE REVIEW

File: tests/test_api.py [hunk 1]
```diff
@@ -43,6 +47,10 @@ def test_status_codes(self, client, endpoint, method, status_code):
         ('/diagnoses/123', 'GET', 'could not find diagnosis `123`'),
         ('/diagnoses/123', 'PUT', 'could not find diagnosis `123`'),
         ('/diagnoses/123', 'DELETE', 'could not find diagnosis `123`'),
+        ('/demographics', 'GET', 'success'),
```
Comment (znatty22) [inline tests/test_api.py]: ... I know I'm terrible...I snuck in a couple missing tests for a couple entities

File: dataservice/api/common/custom_fields.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+)
+
+
+class DateOrDatetime(fields.DateTime):
+    """
+    Custom field that represents a date/datetime field
+    Provide flexible deserialization by allowing values
+    that are either Date or DateTime objects
+
+    * NOTE - Marshmallow's _deserialize method assumes an iso
```
Comment (znatty22) [inline dataservice/api/common/custom_fields.py:15]: See https://github.com/marshmallow-code/marshmallow/blob/3fd59becd91ac95617b0c06ecee2389dfe88dfda/marshmallow/fields.py#L965

This method might have a bug, looks like the `elif utils.dateutil_available` block can never execute?

Comment (dankolbman) [reply dataservice/api/common/custom_fields.py:15]: Looks like `elif self.dateformat` is always true? 
You should submit an issue and offer  a PR here:
https://github.com/marshmallow-code/marshmallow/pulls

Comment (znatty22) [reply dataservice/api/common/custom_fields.py:15]: Yea, I wanted to but didn't have time today. I'll try to get to it tomorrow or the weekend.

Comment (znatty22) [reply dataservice/api/common/custom_fields.py:15]: Aww darn :( I don't know how I missed this but someone already got to it: 

https://github.com/marshmallow-code/marshmallow/pull/619

Comment (dankolbman) [reply dataservice/api/common/custom_fields.py:15]: Should create an issue in our repo to make sure we track this.

File: dataservice/api/common/validation.py [hunk 3]
```diff
@@ -19,3 +19,13 @@ def validate_age(value):
...
         raise ValidationError('Age must be an integer less than {}.'
                               .format(MAX_AGE_DAYS))
+
+
+def validate_positive_number(value):
+    """
+    Validates a that value is a positive number
+    """
+
+    type_str = 'integer' if type(value) == int else 'float'
```
Comment (dankolbman) [inline dataservice/api/common/validation.py]: Could also do `type_str = type(value).__name__`

Does this override the default validator? Does there need to be a check that the value is numeric?

File: tests/test_api.py [hunk 4]
```diff
@@ -141,6 +151,30 @@ def test_relations(self, client, entities, resource, field):
             assert type(body[field]) is str
             assert len(body[field]) == 11
 
+    @pytest.mark.parametrize('method', ['POST', 'PATCH'])
+    @pytest.mark.parametrize('value', [12000,
+                                       '12000',
+                                       'hai der'])
+    @pytest.mark.parametrize('endpoint, field',
+                             [('/aliquots', 'shipment_date')])
+    def test_custom_datetime_field(self, client, entities, endpoint, method,
```
Comment (dankolbman) [inline tests/test_api.py]: May be worth turning this into a general test_bad_input check that takes a method, endpoint, field, and value and sends that.

File: tests/aliquot/test_aliquot_resources.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+
+        # Check error message
+        message = 'could not create aliquot'
+        self.assertIn(message, response['_status']['message'])
+
+        # Check db
+        al = Aliquot.query.first()
+        self.assertIs(al, None)
+
+    # def test_custom_datetime_field(self):
```
Comment (dankolbman) [inline tests/aliquot/test_aliquot_resources.py]: Being tested in test_api?

File: tests/aliquot/test_aliquot_resources.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+    #     message = 'could not create aliquot'
+    #     self.assertIn(message, response['_status']['message'])
+    #     errors = ['positive float', 'positive integer']
+    #     for e in errors:
+    #         self.assertIn(e, response['_status']['message'])
+    #
+    #     # Check db
+    #     self.assertIs(1, Aliquot.query.count())
+
+    def test_post_negative_quantities(self):
```
Comment (dankolbman) [inline tests/aliquot/test_aliquot_resources.py]: Could also be part of an invalid input test in test_api

File: dataservice/api/common/validation.py (hunk 3 above)
Comment (znatty22) [reply dataservice/api/common/validation.py]: Nope, doesn't override the default validator

File: tests/test_api.py (hunk 4 above)
Comment (znatty22) [reply tests/test_api.py]: Oo good idea 👍

File: tests/aliquot/test_aliquot_resources.py (hunk 6 above)
Comment (znatty22) [reply tests/aliquot/test_aliquot_resources.py]: 👍

File: tests/aliquot/test_aliquot_resources.py (hunk 5 above)
Comment (znatty22) [reply tests/aliquot/test_aliquot_resources.py]: Ehh woops 😬

File: tests/conftest.py [hunk 7]
```diff
@@ -63,6 +66,15 @@ def entities(client):
...
             'tumor_descriptor': 'tumor'
         },
+        '/aliquots': {
+            'external_id': 'AL1',
+            'shipment_origin': 'CORIELL',
+            'shipment_destination': 'Baylor',
+            'analyte_type': 'DNA',
+            'concentration': 200,
+            'volume': 13.99,
+            'shipment_date': str(datetime.now().replace(tzinfo=tz.tzutc()))
```
Comment (dankolbman) [inline tests/conftest.py]: If you just want a tz here, you can use `datetime.utcnow()`

**Review by dankolbman — APPROVED** [review]

**Comment by parimalak** [comment]
i think referenced the wrong issue

**Review by parimalak — APPROVED** [review]

**Pull request merged**